client/checkout/classic/new-upe.js

```diff
@@ -0,0 +1,242 @@
+import {
+	PAYMENT_METHOD_NAME_BANCONTACT,
+	PAYMENT_METHOD_NAME_BECS,
+	PAYMENT_METHOD_NAME_CARD,
+	PAYMENT_METHOD_NAME_EPS,
+	PAYMENT_METHOD_NAME_GIROPAY,
+	PAYMENT_METHOD_NAME_IDEAL,
+	PAYMENT_METHOD_NAME_P24,
+	PAYMENT_METHOD_NAME_SEPA,
+	PAYMENT_METHOD_NAME_SOFORT,
+} from '../constants.js';
+import {
+	getSelectedUPEGatewayPaymentMethod,
+	getUpeSettings,
+} from '../utils/upe';
+import showErrorCheckout from '../utils/show-error-checkout';
+import { getAppearance } from '../upe-styles';
+import {
+	appendFingerprintInputToForm,
+	getFingerprint,
+} from '../utils/fingerprint';
+
+jQuery( function ( $ ) {
+	const paymentMethods = getUPEConfig( 'paymentMethodsConfig' );
+	const gatewayUPEComponents = {};
+	const api = new WCPayAPI(
+		{
+			publishableKey: getUPEConfig( 'publishableKey' ),
+			accountId: getUPEConfig( 'accountId' ),
+			forceNetworkSavedCards: getUPEConfig( 'forceNetworkSavedCards' ),
+			locale: getUPEConfig( 'locale' ),
+		},
+		apiRequest
+	);
+
+	let fingerprint = null;
+	let appearance = getUPEConfig( 'upeAppearance' );
+
+	enqueueFraudScripts( getUPEConfig( 'fraudServices' ) );
+	initAppearance();
+	initStripeElements();
+
+	function initAppearance() {
+		if ( ! appearance ) {
+			appearance = getAppearance();
+			api.saveUPEAppearance( appearance );
+		}
+	}
+
+	function initStripeElements() {
+		for ( const paymentMethodType in paymentMethods ) {
+			gatewayUPEComponents[ paymentMethodType ] = {
+				elements: null,
+				upeElement: null,
+			};
+		}
+	}
+
+	/**
+	 * This is the entry-point of this file, which is responsible for mounting the UPE elements.
+	 * It is called when the page is loaded.
+	 *
+	 */
+	$( document.body ).on( 'updated_checkout', () => {
+		$( '.wcpay-upe-element' )
+			.toArray()
+			.forEach( ( domElement ) => {
+				const paymentMethodType = $( domElement ).data(
+					'payment-method-type'
+				);
+				const upeElement =
+					gatewayUPEComponents[ paymentMethodType ].upeElement;
+				if ( upeElement ) {
+					upeElement.mount( domElement );
+				} else {
+					createAndMountPaymentElement(
+						paymentMethodType,
+						domElement
+					);
+				}
+			} );
+	} );
+
+	/**
+	 * Block UI to indicate processing and avoid duplicate submission.
+	 *
+	 * @param {Object} $form The jQuery object for the form.
+	 */
+	const blockUI = ( $form ) => {
+		$form.addClass( 'processing' ).block( {
+			message: null,
+			overlayCSS: {
+				background: '#fff',
+				opacity: 0.6,
+			},
+		} );
+	};
+
+	/**
+	 * Creates the UPE element and mounts it to the DOM element.
+	 * No Stripe PaymentIntent is created at this point.
+	 *
+	 * @param {string} paymentMethodType The type of payment method, e.g. 'card', 'giropay', etc.
+	 * @param {string} domElement The selector of the DOM element of particular payment method to mount the UPE element to.
+	 */
+	async function createAndMountPaymentElement(
+		paymentMethodType,
+		domElement
+	) {
+		const amount = Number( getUPEConfig( 'cartTotal' ) );
```

Comment:
This is a very good observation. There are multiple layers of this. 

#### The amount changes, but the intent type remains to be of type `payment`.
> TL;DR this is covered everywhere.

In this case, legacy UPE/split UPE are handling it but on the back-end, meaning that [the payment intent gets updated](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/payment-methods/class-upe-payment-gateway.php#L239-L256) with the new amount. The Stripe Card Element handles this as well in the right way because it fetches the amount only after the checkout button is clicked, and the amount is proper at that stage. Our new implementation with deferred intent creation handles it because the amount is updated after clicking the _Place order_ button right before the intent is created and confirmed. 

#### The amount changes, implying the intent type change `setup` <-> `payment`.
This is an interesting case, and I think you are mentioning exactly this. I found out that the intent type is not changed in any of the implementations. For the legacy card, things run smoothly since [we don't need to provide the intent type when rendering the UI element](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/classic/index.js#L54-L57). For the legacy UPE and split UPE, the expected behavior is that the available payment methods are updated. If there are eight different gateways for a physical product and it gets removed in the meantime + a subscription with a free trial is added to the cart, and all the gateways except the CC should disappear, which doesn't occur. Eventually, users will see

<img width="796" alt="image" src="https://user-images.githubusercontent.com/22319444/230358158-a787cc6c-afe5-4eaa-b118-a2081950e260.png">
for the split UPE and for the current deferred intent flow implementation

<img width="783" alt="image" src="https://user-images.githubusercontent.com/22319444/230358830-596fde43-9501-44d1-bff5-a20a3ab0f6bb.png">
for the legacy UPE.


The legacy UPE + split UPE should be fixed within a separate issue. I'll make things work for the deferred intent flow. Per my research, there is a `updated_cart_totals` event that is triggered, and we can then [update the element dynamically](https://stripe.com/docs/payments/finalize-payments-on-the-server?platform=web&type=payment#dynamic-updates) which hopefully allows us to successfully change the flow.

Reply:
@FangedParakeet There are a few things to consider, and I'd like to hear your thoughts on what and how we would like to prioritize here.

#### No third-party plugins
Without third-party plugins, this can break as described above for any payment gateways (legacy UPE/split UPE etc.). And to make this happen, one needs to update the cart in a separate browser tab and then place an order on the initial checkout page in another tab. Given the nature of PHP and how it's rolled out, there is no easy way to fix this, and I don't think we should search for ways to update the rolled-out scripts in a separate tab. As you rightly pointed out yesterday, one good option here would be to polish the error message and make it clearer for users. As discussed yesterday, I will create a separate issue soon.

#### Third-party plugins
Regarding third-party plugins that allow modifying the total on the checkout page. It looks like`updated_cart_totals` DOM event should be triggered after such changes, which makes it possible for us to handle the event and [dynamically update the Stripe Elements](https://stripe.com/docs/payments/finalize-payments-on-the-server?platform=web&type=payment#dynamic-updates) whenever the change occurs. However, there are two concerns I have in mind: 

1. I'm not sure we want to add the support logic for the third-party plugins in the main gateway script that handles payment processing for WooCommerce Payments, and even if we want, I think this could be done as part of the issue I mentioned above for all the other UPE types since the problem isn't local. 
2. Those gateways that can't be used for subscription purchasing will disappear after the Stripe Element for them is updated. Then, we will have an empty `div` (see below) and need to find a way to remove this without refreshing the page.
<img width="327" alt="image" src="https://user-images.githubusercontent.com/22319444/230621831-727d04bb-965f-407b-b6e2-c01bd6540d3f.png">


All this, while not impossible, feels like a new feature for the existing payments, which makes me feel we might want to brainstorm ideas and address this defect across all the UPEs. Please let me know what you think about all this - I'm happy to provide more information in case needed.

Reply:
>While not impossible, all this feels like a new feature for the existing payments, making me feel we might want to brainstorm ideas and address this defect across all the UPEs separately.

For now, I'd rather focus on achieving parity with the legacy UPE rather than comprehensively perfecting our implementation. So in any scenario, I think it's fine for us to return an error, if the existing UPE implementation also returns an error; but if the legacy implementation has some mystical manner of handling such edge cases, we will need to devise one ourselves as well. 

> In order to change the product list completely and move from a physical product to a subscription with a free trial, users still need to go back to the cart page.

The only situation I can think would be if a user had two items in their cart: a regular simple product and a free-trial subscription product. At this point, presumably we'd complete the checkout using a payment intent; however, if a coupon were added to remove the simple product, reducing the checkout total to zero, we would still need to collect payment for the subscription product, but we would now need a setup intent and would not charge this payment method at this time. I will need to test to see how the legacy implementation handles this case, but--if it is even a scenario we will need to entertain--it is definitely an edge case that can be resolved in a separate issue.

Reply:
> if a user had two items in their cart: a regular simple product and a free-trial subscription product. At this point, presumably we'd complete the checkout using a payment intent; however, if a coupon were added to remove the simple product, reducing the checkout total to zero, we would still need to collect payment for the subscription product, but we would now need a setup intent and would not charge this payment method at this time. 

Legacy UPE and split UPE are failing here. Legacy card and deferred intent creation flow handle this scenario. The fact that the intent is created after order placement seems to play a key role, and the Stripe UI Element's `mode` is not relevant in this use-case.